### PR TITLE
Toyota: LTA steering control safety

### PR DESCRIPTION
**Pull request by sshane**

### CODE REVIEW

File: board/safety/safety_toyota.h [hunk 1]
```diff
@@ -68,6 +84,21 @@ static uint32_t toyota_get_checksum(CANPacket_t *to_push) {
...
+  int addr = GET_ADDR(to_push);
+
+  bool valid = false;
+  if (toyota_lta) {
+    if (addr == 0x260) {
+      // This is always 1 on non-TSS2 platforms
+      valid = GET_BIT(to_push, 3U) == 0U;  // STEER_ANGLE_INITIALIZING
+    }
+  } else {
+    valid = true;
```
Comment (adeebshihadeh) [inline board/safety/safety_toyota.h:97]: remove this, default should be false

File: board/safety/safety_toyota.h [hunk 2]
```diff
@@ -175,19 +229,51 @@ static bool toyota_tx_hook(CANPacket_t *to_send) {
...
+        // check angle rate limits and inactive angle
+        if (steer_angle_cmd_checks(lta_angle, steer_control_enabled, TOYOTA_STEERING_LIMITS)) {
+          tx = false;
+        }
+
+        if (lta_request != lta_request2) {
+          tx = false;
+        }
+
+        // SETME_X64 is gated on steer request
```
Comment (adeebshihadeh) [inline board/safety/safety_toyota.h:257]: what is this? can we give it a better name?

File: board/safety/safety_toyota.h [hunk 3]
```diff
@@ -68,6 +84,21 @@ static uint32_t toyota_get_checksum(CANPacket_t *to_push) {
...
 }
 
+static bool toyota_get_quality_flag_valid(CANPacket_t *to_push) {
+  int addr = GET_ADDR(to_push);
+
+  bool valid = false;
+  if (toyota_lta) {
+    if (addr == 0x260) {
+      // This is always 1 on non-TSS2 platforms
+      valid = GET_BIT(to_push, 3U) == 0U;  // STEER_ANGLE_INITIALIZING
```
Comment (adeebshihadeh) [inline board/safety/safety_toyota.h:94]: what's this for?

File: board/safety/safety_toyota.h [hunk 4]
```diff
@@ -101,10 +146,19 @@ static void toyota_rx_hook(CANPacket_t *to_push) {
...
+      int speed = 0;
+      // sum 4 wheel speeds. conversion: raw * 0.01 - 67.67
+      for (uint8_t i = 0U; i < 8U; i += 2U) {
+        int wheel_speed = (GET_BYTE(to_push, i) << 8U) + GET_BYTE(to_push, (i + 1U));
+        speed += wheel_speed - 0x1a6f;
+      }
+      // check that all wheel speeds are at zero value
+      vehicle_moving = speed != 0;
+
+      speed = ROUND(speed / 4.0 * 0.01 / 3.6 * VEHICLE_SPEED_FACTOR);
```
Comment (adeebshihadeh) [inline board/safety/safety_toyota.h:160]: this isn't super readable

File: board/safety/safety_toyota.h (hunk 3 above)
Comment (sshane) [reply board/safety/safety_toyota.h:94]: This is a quality flag for the current angle measurement, which we only use to check that the desired angle matches current angle while inactive.

File: board/safety/safety_toyota.h (hunk 2 above)
Comment (sshane) [reply board/safety/safety_toyota.h:260]: https://github.com/commaai/opendbc/pull/983

File: board/safety_declarations.h [hunk 5]
```diff
@@ -220,7 +221,7 @@ bool brake_pressed_prev = false;
 bool regen_braking = false;
 bool regen_braking_prev = false;
 bool cruise_engaged_prev = false;
-struct sample_t vehicle_speed;
+struct sample_t vehicle_speed;  // m/s
```
Comment (sshane) [inline board/safety_declarations.h:224]:
```suggestion
struct sample_t vehicle_speed;
```

File: board/safety/safety_ford.h [hunk 6]
```diff
@@ -214,7 +214,7 @@ static void ford_rx_hook(CANPacket_t *to_push) {
     // Update vehicle speed
     if (addr == FORD_BrakeSysFeatures) {
       // Signal: Veh_V_ActlBrk
-      update_sample(&vehicle_speed, ROUND(((GET_BYTE(to_push, 0) << 8) | GET_BYTE(to_push, 1)) * 0.01 / 3.6 * VEHICLE_SPEED_FACTOR));
+      UPDATE_VEHICLE_SPEED(((GET_BYTE(to_push, 0) << 8) | GET_BYTE(to_push, 1)) * 0.01 / 3.6);
```
Comment (adeebshihadeh) [inline board/safety/safety_ford.h:217]: what's a change to safety_ford doing here?

File: board/safety/safety_toyota.h (hunk 4 above)
Comment (sshane) [reply board/safety/safety_toyota.h]: https://github.com/commaai/panda/pull/1762

**Pull request merged**